lib/recursive-methods.rb

```diff
@@ -1,49 +1,103 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) as n gets bigger, you make that many recursive calls
+# Space complexity: O(n) due to taking up space on the system call stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+
+  raise ArgumentError, "Value must be greater than or equal to 0" if n < 0
+
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2). s[1..-1] creates a new string
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+
+  rev_str = reverse(s[1..-1])
+  rev_str += s[0]
+  rev_str
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n) n is the length of the string s
+# Space complexity: O(n) due to taking up space on the system call stack
+def reverse_inplace(s, left_indx = 0, right_indx = s.length - 1)
+  return s if left_indx >= right_indx
+  # multiple assignment where the left and right indexes switch positions in the string
+  s[left_indx], s[right_indx] = s[right_indx], s[left_indx]
+
+  return reverse_inplace(s, left_indx + 1, right_indx - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) n is the number of elements
+# Space complexity: O(n) taking up space on the system call stack depending on the size of n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n) n is th length of the string s
+# Space complexity: O(n) dependent on the size of n (string)
+def nested(s, left = "(", right = ")")
+  return true if s.empty?
+  return false if s.count(left) != s.count(right)
```

Comment:
This is going to add O(n) time complexity with each iteration... so O(n<sup>2</sup>)